build: fix accidental cmake expansions

stephank opened the pull request:
As of CMake 3.25, there are now global variables `LINUX=1`, `ANDROID=1`, etc. These conflict with expressions that used these names as unquoted strings in positions where CMake accepts 'variable|string', for example:

- `if(sdk STREQUAL LINUX)` would fail, because `LINUX` is now defined and expands to 1, where it would previously coerce to a string.

- `if(${sdk} STREQUAL "LINUX")` would fail if `sdk=LINUX`, because the left-hand side expands twice.

In this patch, I looked for a number of patterns to fix up, sometimes a little defensively:

- Quoted right-hand side of `STREQUAL` where I was confident it was intended to be a string literal.

- Removed manual variable expansion on left-hand side of `STREQUAL`, `MATCHES` and `IN_LIST` where I was confident it was unintended.

Fixes #65028.

---

Question: I have [a branch for 5.8](https://github.com/stephank/swift/tree/fix/5.8-cmake-3.25). This is what I've actually been testing [in Nixpkgs](https://github.com/NixOS/nixpkgs/pull/225197), and that builds on x86-64 Linux and aarch64 macOS. Between 5.8 and `main`, it looks like the only conflicts were CMake additions in `main`. Should I open a PR against `release/5.8` now, or after review?

stephank committed:
build: fix accidental cmake expansions

As of CMake 3.25, there are now global variables `LINUX=1`, `ANDROID=1`,
etc. These conflict with expressions that used these names as unquoted
strings in positions where CMake accepts 'variable|string', for example:

- `if(sdk STREQUAL LINUX)` would fail, because `LINUX` is now defined and
  expands to 1, where it would previously coerce to a string.

- `if(${sdk} STREQUAL "LINUX")` would fail if `sdk=LINUX`, because the
  left-hand side expands twice.

In this patch, I looked for a number of patterns to fix up, sometimes a
little defensively:

- Quoted right-hand side of `STREQUAL` where I was confident it was
  intended to be a string literal.

- Removed manual variable expansion on left-hand side of `STREQUAL`,
  `MATCHES` and `IN_LIST` where I was confident it was unintended.

Fixes #65028.

buttaface commented:
> Should I open a PR against release/5.8 now, or after review?

The way it usually works is that you get it into `main` first and if nothing breaks, you can cherry-pick it for the release branches and submit pulls for those, eg this recent 5.8 pull #63995.

egorzhdan commented:
@swift-ci please smoke test

tokyovigilante commented:
Thanks, can confirm this significantly improves build progress on Fedora 39 (rawhide), however I still see the following building Swift:

```
240/1501][ 15%][17.157s] Building CXX object lib/Demangling/CMakeFiles/swiftDemangling.dir/ManglingUtils.cpp.o
FAILED: lib/Demangling/CMakeFiles/swiftDemangling.dir/ManglingUtils.cpp.o
/usr/lib64/ccache/clang++ -DCMARK_STATIC_DEFINE -DGTEST_HAS_RTTI=0 -DLLVM_DISABLE_ABI_BREAKING_CHECKS_ENFORCING=1 -DSWIFT_STDLIB_HAS_TYPE_PRINTING=1 -DSWIFT_SUPPORT_OLD_MANGLING=1 -DSWIFT_THREADING_LINUX -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_LIBCPP_ENABLE_ASSERTIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/ryan/Projects/Develop/swift-source/build/buildbot_linux/swift-linux-x86_64/lib/Demangling -I/home/ryan/Projects/Develop/swift-source/swift/lib/Demangling -I/home/ryan/Projects/Develop/swift-source/swift/stdlib/public/SwiftShims -I/home/ryan/Projects/Develop/swift-source/build/buildbot_linux/swift-linux-x86_64/include -I/home/ryan/Projects/Develop/swift-source/swift/include -I/home/ryan/Projects/Develop/swift-source/llvm-project/llvm/include -I/home/ryan/Projects/Develop/swift-source/build/buildbot_linux/llvm-linux-x86_64/include -I/home/ryan/Projects/Develop/swift-source/llvm-project/clang/include -I/home/ryan/Projects/Develop/swift-source/build/buildbot_linux/llvm-linux-x86_64/tools/clang/include -I/home/ryan/Projects/Develop/swift-source/cmark/src/include -I/home/ryan/Projects/Develop/swift-source/build/buildbot_linux/cmark-linux-x86_64/src -I/home/ryan/Projects/Develop/swift-source/swift-corelibs-libdispatch/src/BlocksRuntime -I/home/ryan/Projects/Develop/swift-source/swift-corelibs-libdispatch -Wno-unknown-warning-option -Werror=unguarded-availability-new -fno-stack-protector -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-class-memaccess -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -Werror=switch -Wdocumentation -Wimplicit-fallthrough -Wunreachable-code -Woverloaded-virtual -DOBJC_OLD_DISPATCH_PROTOTYPES=0 -O3 -DNDEBUG -std=c++17  -fno-exceptions -funwind-tables -fno-rtti -Werror=gnu -Werror=c++98-compat-extra-semi -UNDEBUG -target x86_64-unknown-linux-gnu -O2 -g0 -march=core2 -MD -MT lib/Demangling/CMakeFiles/swiftDemangling.dir/ManglingUtils.cpp.o -MF lib/Demangling/CMakeFiles/swiftDemangling.dir/ManglingUtils.cpp.o.d -o lib/Demangling/CMakeFiles/swiftDemangling.dir/ManglingUtils.cpp.o -c /home/ryan/Projects/Develop/swift-source/swift/lib/Demangling/ManglingUtils.cpp
/home/ryan/Projects/Develop/swift-source/swift/lib/Demangling/ManglingUtils.cpp:83:16: error: no member named 'None' in namespace 'llvm'
  return llvm::None;
         ~~~~~~^
1 error generated.
[241/1501][ 16%][17.294s] Building CXX object lib/Demangling/CMakeFiles/swiftDemangling.dir/Demangler.cpp.o
FAILED: lib/Demangling/CMakeFiles/swiftDemangling.dir/Demangler.cpp.o
/usr/lib64/ccache/clang++ -DCMARK_STATIC_DEFINE -DGTEST_HAS_RTTI=0 -DLLVM_DISABLE_ABI_BREAKING_CHECKS_ENFORCING=1 -DSWIFT_STDLIB_HAS_TYPE_PRINTING=1 -DSWIFT_SUPPORT_OLD_MANGLING=1 -DSWIFT_THREADING_LINUX -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_LIBCPP_ENABLE_ASSERTIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/ryan/Projects/Develop/swift-source/build/buildbot_linux/swift-linux-x86_64/lib/Demangling -I/home/ryan/Projects/Develop/swift-source/swift/lib/Demangling -I/home/ryan/Projects/Develop/swift-source/swift/stdlib/public/SwiftShims -I/home/ryan/Projects/Develop/swift-source/build/buildbot_linux/swift-linux-x86_64/include -I/home/ryan/Projects/Develop/swift-source/swift/include -I/home/ryan/Projects/Develop/swift-source/llvm-project/llvm/include -I/home/ryan/Projects/Develop/swift-source/build/buildbot_linux/llvm-linux-x86_64/include -I/home/ryan/Projects/Develop/swift-source/llvm-project/clang/include -I/home/ryan/Projects/Develop/swift-source/build/buildbot_linux/llvm-linux-x86_64/tools/clang/include -I/home/ryan/Projects/Develop/swift-source/cmark/src/include -I/home/ryan/Projects/Develop/swift-source/build/buildbot_linux/cmark-linux-x86_64/src -I/home/ryan/Projects/Develop/swift-source/swift-corelibs-libdispatch/src/BlocksRuntime -I/home/ryan/Projects/Develop/swift-source/swift-corelibs-libdispatch -Wno-unknown-warning-option -Werror=unguarded-availability-new -fno-stack-protector -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-class-memaccess -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -Werror=switch -Wdocumentation -Wimplicit-fallthrough -Wunreachable-code -Woverloaded-virtual -DOBJC_OLD_DISPATCH_PROTOTYPES=0 -O3 -DNDEBUG -std=c++17  -fno-exceptions -funwind-tables -fno-rtti -Werror=gnu -Werror=c++98-compat-extra-semi -UNDEBUG -target x86_64-unknown-linux-gnu -O2 -g0 -march=core2 -MD -MT lib/Demangling/CMakeFiles/swiftDemangling.dir/Demangler.cpp.o -MF lib/Demangling/CMakeFiles/swiftDemangling.dir/Demangler.cpp.o.d -o lib/Demangling/CMakeFiles/swiftDemangling.dir/Demangler.cpp.o -c /home/ryan/Projects/Develop/swift-source/swift/lib/Demangling/Demangler.cpp
In file included from /home/ryan/Projects/Develop/swift-source/swift/lib/Demangling/Demangler.cpp:23:
In file included from /home/ryan/Projects/Develop/swift-source/swift/include/swift/Strings.h:16:
/home/ryan/Projects/Develop/swift-source/swift/include/swift/Basic/LLVM.h:65:30: error: redefinition of 'Optional' as different kind of symbol
  template<typename T> class Optional;
                             ^
/home/ryan/Projects/Develop/swift-source/llvm-project/llvm/include/llvm/ADT/Optional.h:24:20: note: previous definition is here
template <class T> using Optional = std::optional<T>;
                   ^
1 error generated.
[247/1501][ 16%][21.886s] Building CXX object stdlib/public/RemoteInspection/CMakeFiles/swiftRemoteInspection-linux-x86_64.dir/TypeRefBuilder.cpp.o
ninja: build stopped: subcommand failed.
ERROR: command terminated with a non-zero exit status 1, aborting

ERROR: command terminated with a non-zero exit status 1, aborting
```

stephank commented:
It's possible I missed some detail on the `main` branch. I built this change on `main` without testing, then backported to 5.8 and tested that, because my intent was to package 5.8 for Nixpkgs.

I'm currently not setup to easily test other branches of Swift than the one version we package in Nixpkgs. Adding some support for that to Nixpkgs is something I'd like, but also a bit of work. I can't really plan that work, because this is spare time effort for me. :)

Perhaps a follow-up PR can fix further issues?

tokyovigilante commented:
Ok, thanks, I appreciate your efforts anyway and good to have some idea of the root cause. The above is against `-next` so I will try a `-main` build also.

bnbarham commented:
> Ok, thanks, I appreciate your efforts anyway and good to have some idea of the root cause. The above is against -next so I will try a -main build also.

`next` will definitely not build. We branch `next` onto `rebranch` periodically and fix up the new LLVM API at that point based on a `stable/*` branch in the apple LLVM fork.

tokyovigilante commented:
Just noting that changes in https://github.com/apple/swift/commit/b5866040ae7348515beebcc4e3cda167781439a4 mean that the changes in `stdlib/public/Cxx/std/CMakeLists.txt` now need to be made to `stdlib/public/Cxx/libstdcxx/CMakeLists.txt`. Have retested more recent `main`(https://github.com/apple/swift/commit/eb76d9afe77b24f4326742480589451ad1a655e5) and build succeeds on Fedora with this change.

finagolfin commented:
@stephank, please rebase.

@compnerd, this appears to work well, fixing a recent CMake issue on linux, mind reviewing?

compnerd commented on cmake/modules/AddSwift.cmake:
```diff
@@ -427,7 +427,7 @@ function(_add_host_variant_link_flags target)
   # TODO: Evaluate/enable -f{function,data}-sections --gc-sections for bfd,
   # gold, and lld.
-  if(NOT CMAKE_BUILD_TYPE STREQUAL Debug AND CMAKE_SYSTEM_NAME MATCHES Darwin)
+  if(NOT CMAKE_BUILD_TYPE STREQUAL "Debug" AND CMAKE_SYSTEM_NAME MATCHES Darwin)
```
Should the RHS of the `MATCHES` be quoted as well?  While a regex, it is still stringly-typed right?

stephank replied:
I've not seen any breakage because of this in our Nixpkgs build. `MATCHES` RHS doesn't accept `variable`, so should be safe, I believe.

However, if you'd like me to fix this up, it's easy enough to grep. There are only 18 occurrences. (And FWIW, no cases of `LINUX` or `ANDROID`.)

finagolfin commented:
@stephank, this cannot go anywhere till you rebase and fix any conflicts.

stephank commented:
Thanks for the reminder. I've been away for a bit, but took another look just now.

The conflict was a simple move of a CMake code snippet in b5866040ae7348515beebcc4e3cda167781439a4, so easy to resolve.

I diff'd between merge base (404b925d8b6c1f323900ea765132924150b1e54b) and `main` at the time of writing (97e6b56456f31bb7009e9310f93c80bd98862f47), and also added some new changes in line with the rest of this PR:

- https://github.com/apple/swift/pull/65534/files#diff-fe0eb0f4f578c2c4b5ef562e9ac0dfacd3c037901114f34f7a90a8a6d42fc7e3L144-R144
- https://github.com/apple/swift/pull/65534/files#diff-fe0eb0f4f578c2c4b5ef562e9ac0dfacd3c037901114f34f7a90a8a6d42fc7e3L270-R270
- https://github.com/apple/swift/pull/65534/files#diff-4b74006b7c38d950627c4183690f1c6b5435d1b4da7a327fdf6ed9bcbb2cc33bL936-R936
- https://github.com/apple/swift/pull/65534/files#diff-5c1eb2d6768a23753402274997290de28b4c44906856c32d858730083df6857eL250-R250
- https://github.com/apple/swift/pull/65534/files#diff-1c94ef1e032a4915d67773ad4157fec0c958d4e45603be517d416fa8b5f62a34L162-R162

finagolfin commented:
Ping @etcwilde, would you take a look?

finagolfin commented:
@stephank, I think another rebase is in order.

Pinging @etcwilde again, Saleem signed off but would like you to take a look.

stephank commented:
The rebase was trivial, and I did not find new issues at a glance diffing 97e6b56456f..c50263c7301.

finagolfin commented:
Ping @edymtt, maybe you have a moment to review this.

finagolfin commented:
@bnbarham, one final CI run and we can get this in.

bnbarham commented:
@swift-ci please test

finagolfin commented:
The Mac CI alone complains that it can't find the Swift module for iOS:
```
<unknown>:0: error: could not find module 'Swift' for target 'arm64-apple-ios'; found: x86_64-apple-ios-simulator, arm64-apple-ios-simulator, at: /Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/swift-macosx-x86_64/lib/swift/iphonesimulator/Swift.swiftmodule
error: fatalError
```
I don't know if that's because of the couple iOS modifications in this pull or if this could be a spurious CI error.

finagolfin commented:
@bnbarham, would you run the mac CI on this again? Let's see if the last failure reproduces, as I suspect it was spurious.

bnbarham commented:
I don't think I've seen that one (and I'm wrangling this week). But I'll re-run 👍

bnbarham commented:
@swift-ci please test macOS platform

finagolfin commented:
OK, it reproduces.

@stephank, any idea why this pull would be failing like this on macOS alone?

stephank commented:
I've been staring at the PR, but can't find my mistake. It seems the Jenkins log has expired. I'll try to find some time to reproduce it soon.

finagolfin commented:
@kateinoigakukun, would you run the mac CI on this again, so he can examine the log, as the prior log has been removed?

kateinoigakukun commented:
@swift-ci please test macOS platform

finagolfin commented:
I don't think it's setting the Apple targets the way it should. I diffed a recent passing build log against this one and see two mistakes in the CMake config, both related to the simulators.

The passing one says:
```
-- iOS Simulator SDK:
--   Object File Format: MACHO
--   Swift Standard Library Path: iphonesimulator
--   Threading Package: darwin
--   Version: 16.2
--   Build number: 20C52
--   Deployment version: 11.0
--   Triple name: ios
--   Simulator: TRUE
--   Architectures: x86_64;arm64
--
-- Assuming watchOS Simulator SDK at /Applications/Xcode.app/Contents/Developer/Platforms/WatchSimulator.platform/Developer/SDKs/WatchSimulator9.1.sdk supports architecture i386
-- watchOS Simulator SDK:
--   Object File Format: MACHO
--   Swift Standard Library Path: watchsimulator
--   Threading Package: darwin
--   Version: 9.1
--   Build number: 20S71
--   Deployment version: 4.0
--   Triple name: watchos
--   Simulator: TRUE
```
while this pull strangely produces this, ie both simulator SDKs are not listed as simulators:
```
-- iOS Simulator SDK:
--   Object File Format: MACHO
--   Swift Standard Library Path: iphonesimulator
--   Threading Package: darwin
--   Version: 16.2
--   Build number: 20C52
--   Deployment version: 11.0
--   Triple name: ios
--   Simulator: FALSE
--   Architectures: x86_64;arm64
--
-- Assuming watchOS Simulator SDK at /Applications/Xcode.app/Contents/Developer/Platforms/WatchSimulator.platform/Developer/SDKs/WatchSimulator9.1.sdk supports architecture i386
-- watchOS Simulator SDK:
--   Object File Format: MACHO
--   Swift Standard Library Path: watchsimulator
--   Threading Package: darwin
--   Version: 9.1
--   Build number: 20S71
--   Deployment version: 4.0
--   Triple name: watchos
--   Simulator: FALSE
```

stephank commented:
Right, it looks like `macro` arguments are interpolated differently, and are not actually variables. They can't be used in `variable` position:

```
cmake_minimum_required(VERSION 3.10)
project(cmaktest)

macro(foo name)
  message("Testing: ${name}")

  if(name MATCHES "simulator")
    message("Y")
  else()
    message("N")
  endif()
endmacro()

foo(watchsimulator)
```

Prints:

```
Testing: watchsimulator
N
```

Fixes are:
- `if("${name}" MATCHES ...`
- Changing it to a function
- Creating an actual variable (`set(name ...`) before the macro call

The `xcrun_name` use was the only instance I could find where this happens. I changed that to a quoted expansion, and rebased.

etcwilde commented:
> Right, it looks like macro arguments are interpolated differently, and are not actually variables.

Oh yeah! I'd forgotten about that. Another reason why I usually like to steer away from macros. Thanks for looking into this further.

finagolfin commented:
@bnbarham, hopefully one last CI run?

bnbarham commented:
@swift-ci please test

finagolfin commented:
Sourcekit-lsp test on linux CI timed out, appears completely unrelated, that CI alone will need to be rerun.

etcwilde commented:
@swift-ci please test Linux

stephank commented:
I think that's the same sourcekit-lsp timeout error again?

bnbarham commented:
@swift-ci please test Linux

bnbarham commented:
> I think that's the same sourcekit-lsp timeout error again?

Yes, the revert wasn't merged until this morning.

finagolfin commented:
@al45tair, ready for merge.

finagolfin commented:
@etcwilde, ready for merge.

finagolfin commented:
@stephank, do you want to submit this for 5.9 next? See [the 5.9 release process](https://forums.swift.org/t/swift-5-9-release-process/63557) and a recent example pull, #67391.

RemiBardon commented:
@stephank Do you have any plan on backporting to `release/5.8`? I'm having the issue as well on [Exherbo](https://www.exherbo.org)

stephank commented:
> @stephank Do you have any plan on backporting to `release/5.8`? I'm having the issue as well on [Exherbo](https://www.exherbo.org)

I don't have a need for it myself, so was not planning on doing it. Is there even going to be another 5.8 release?

This is the patch Nixpkgs currently uses for 5.8: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/swift/compiler/patches/swift-cmake-3.25-compat.patch

finagolfin commented:
> Is there even going to be another 5.8 release?

There might: the fall release like 5.9 usually releases in September, and there is sometimes an August patch release before that.

@etcwilde, wdyt, should he submit this for 5.8 too?
